students/cfarris/lessons/lesson09/assignment/mdb_db.py

```diff
@@ -0,0 +1,217 @@
+
+
+def print_mdb_collection(db, collection_name):
+    """Prints the collection specified"""
+
+    for doc in collection_name.find():
+        print(doc)
+
+
+def import_data(db, customers=None, products=None, rentals=None, start_time=0):
+    """
+    Adds files to collection.
+    if the file/date type (k) fits into three categories
+    it will be added to that collection.
+    I created a thread for rentals as well,
+    but commented it out to adhere to instructions.
+    """
+
+    data_list = []
+
+    # if statements were added to facilitate testing.
+    if products:
+        products_db = db['products_db']
+        _read_csv(db, products_db, products, data_list, start_time)
+
+    if customers:
+        customers_db = db['customers_db']
+        _read_csv(db, customers_db, customers, data_list, start_time)
+
+    if rentals:
+        rentals_db = db['rentals_db']
+        _read_csv(db, rentals_db, rentals, data_list, start_time)
+
+    print("data_list: ", data_list)
+    return data_list
+
+
+def _read_csv(db, collection, csv_file, data_list=None, start_time=0):
+    '''
+    Opens csv files, creates a list of
+    dictionarys with matching keys from header to values
+    returns the dictionary and saves it to
+    different mongodb collections based upon the
+    file name given to main.
+    '''
+
+    if data_list is None:
+        data_list = []
+
+    with open(csv_file, 'r', encoding='utf-8-sig') as csv_file:
+        csv_reader = csv.reader(csv_file, delimiter=',')
+        header = True
+        first_line = []
+        num_processed = 0
+        coll_of_dicts = []
+        count_before_running = collection.count()
+        for row in csv_reader:
+            if header:
+                header = False
+                first_line = [value.strip() for value in row]
+            else:
+                num_processed += 1
+                new_dict = {}
+                str_row = []
+                str_row = [value.strip() for value in row]
+                for i in range(len(str_row)):
+                    k = first_line[i]
+                    v = str_row[i]
+                    novelty_test = check_redundancy(db, k, v, collection)
+                    if novelty_test:
+                        new_dict[first_line[i]] = str_row[i]
+                coll_of_dicts.append(new_dict)
+        try:
+            collection.insert_many(coll_of_dicts)
+        except Exception as err:
+            print("error!:", err, collection.name, coll_of_dicts)
+        time_to_run = time.time() - start_time
+        results = (num_processed,
+                   count_before_running,
+                   collection.count(),
+                   time_to_run)
+        print(collection.name,
+              " finished: ",
+              num_processed,
+              count_before_running,
+              collection.count(),
+              time_to_run)
+        data_list.append(results)
+        return data_list
```

Comment:
There is way too much in this context manager, even the return statement is in the context manager. Try to do just the minimum, and then close the connection. There is also too much in this function. Notice that your doc string says this function is doing two things, returning a dictionary and saving data, so this should probably be two functions. Sorry if I am repeating myself, never sure what I've said in past reviews. 🤷‍♀️